File: problem/Min_ASCII_Del_Sum/minimumDeleteSum.js
/**
 * @param {string} s1
 * @param {string} s2
 * @return {number}
 */
var minimumDeleteSum = function(s1, s2) {
    const n = s1.length, m = s2.length;
    const dp = Array.from({ length: n + 1 }, () => Array(m + 1).fill(0));

    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            if (s1[i] === s2[j])
                dp[i + 1][j + 1] = dp[i][j] + s1.charCodeAt(i);
            else
                dp[i + 1][j + 1] = Math.max(dp[i][j + 1], dp[i + 1][j]);
        }
    }

    let total = 0;
    for (const c of s1) total += c.charCodeAt(0);
    for (const c of s2) total += c.charCodeAt(0);

    return total - 2 * dp[n][m];
};

File: problem/Min_ASCII_Del_Sum/minimumDeleteSum.c
int max(int a, int b)
{
    return a > b ? a : b;
}

int minimumDeleteSum(char *s1, char *s2)
{
    int n = strlen(s1);
    int m = strlen(s2);

    int **dp = (int **)malloc((n + 1) * sizeof(int *));
    for (int i = 0; i <= n; i++)
    {
        dp[i] = (int *)calloc(m + 1, sizeof(int));
    }

    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < m; j++)
        {
            if (s1[i] == s2[j])
                dp[i + 1][j + 1] = dp[i][j] + (int)s1[i];
            else
                dp[i + 1][j + 1] = max(dp[i][j + 1], dp[i + 1][j]);
        }
    }

    int total = 0;
    for (int i = 0; i < n; i++)
        total += s1[i];
    for (int j = 0; j < m; j++)
        total += s2[j];

    int ans = total - 2 * dp[n][m];

    for (int i = 0; i <= n; i++)
        free(dp[i]);
    free(dp);

    return ans;
}